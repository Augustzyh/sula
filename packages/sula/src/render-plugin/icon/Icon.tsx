import React from 'react';
import cx from 'classnames';
import warning from '../../_util/warning';
import './style/icon.less';

type BasicIconProps = {
  style?: React.CSSProperties;
  className?: string;
  spin?: boolean;
  rotate?: number;
  twoToneColor?: string;
};

export type IconThemeMapper = {
  outlined?: React.ComponentClass<BasicIconProps>;
  filled?: React.ComponentClass<BasicIconProps>;
  twoTone?: React.ComponentClass<BasicIconProps>;
};

export type IconMapper = Record<string, IconThemeMapper>;

type Props = {
  type: string;
  theme: 'filled' | 'twoTone';

  disabled?: boolean;
  text?: string | React.ReactElement;
  iconMapper?: IconMapper; // 独立使用
} & BasicIconProps;

type NativeAttrs = Omit<React.HTMLAttributes<any>, keyof Props>;

export type IconProps = Props & NativeAttrs;

let globalIconMapper: IconMapper = {};

export default class Icon extends React.Component<IconProps> {
  static defaultProps = {
    theme: 'outlined',
  };

  static iconRegister = (iconMapper: IconMapper) => {
    globalIconMapper = iconMapper;
  };

  render() {
    const { type, theme, disabled, text, iconMapper, ...iconProps } = this.props;
    const finalIconMapper: IconMapper = iconMapper || globalIconMapper;
    if (finalIconMapper[type] && finalIconMapper[type][theme]) {
      const iconCls = cx(iconProps.className, {
        'sula-icon-disabled': !!disabled,
        'sula-icon-clickable': !!iconProps.onClick && disabled !== true,
      });

      const IconCompClass = finalIconMapper[type][theme] as React.ComponentClass<BasicIconProps>;
      if (text) {
        const { rotate, twoToneColor, spin, className, ...restProps } = iconProps;
        return (
          <span className={iconCls} {...restProps}>
            {React.createElement(IconCompClass, {
              rotate,
              spin,
              twoToneColor,
            })}
            <span style={{ marginLeft: 4 }}>{text}</span>
          </span>
        );
      } else {
        iconProps.className = iconCls;
        return React.createElement(IconCompClass, iconProps);
      }
    }

    warning(false, 'icon', `${type} ${theme} is not registered in iconMapper`);
    return null;
  }
}
