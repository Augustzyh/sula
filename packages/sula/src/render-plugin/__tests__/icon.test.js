import React from 'react';
import { TabletFilled, AppstoreOutlined } from '@ant-design/icons';
import { mount, shallow } from 'enzyme';
import Icon from '../icon/Icon';
import LoadingIcon from '../icon/InnerLoadingIcon';
import { IconPlugin } from '../icon';
import '../../__tests__/common';

describe('icon', () => {
  jest.useFakeTimers(); // 设置快速时间
  beforeEach(() => {
    Icon.iconRegister({
      tablet: {
        filled: TabletFilled,
      },
      appstore: {
        outlined: AppstoreOutlined,
      }
    })
  });

  describe('Icon test', () => {
    it('basic icon', () => {
      const wrapper = mount(<Icon type="Head" />);
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.unmount();
    })
  
    it('no text', () => {
      const wrapper = mount(<Icon type="tablet" theme="filled" className="testClassName" />);
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.unmount();
    })
  
    it('has text', () => {
      const wrapper = mount(<Icon type="tablet" theme="filled" text="testIcon" onClick={() => {}} disabled={false} />);
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.unmount();
    })
  
    it('icon click', () => {
      class TestIcon extends React.Component{
        state = {
          iconText: 'iconText',
          disabled: false,
        }
  
        render() {
          const { iconText, disabled } = this.state;
          return (
            <Icon
              disabled
              type="tablet"
              theme="filled"
              text={iconText}
              onClick={() => {
                this.setState({
                  iconText: iconText ? undefined : 'iconText',
                  disabled: !disabled
                })
              }}
            />
          )
        }
      }
      const wrapper = shallow(<TestIcon />);
      expect(wrapper.state('iconText')).toBe('iconText');
      expect(wrapper.state('disabled')).toBe(false);
      wrapper.find(Icon).simulate('click');
      expect(wrapper.state('iconText')).toBe(undefined);
      expect(wrapper.state('disabled')).toBe(true);
    })
  })

  describe('LoadingIcon', () => {
    it('basic loadingIcon', () => {
      const wrapper = mount(<LoadingIcon />);
      expect(wrapper.render()).toMatchSnapshot();
    })
  })

  describe('LoadingIconManager', () => {
    it('autoLoading is false', () => {
      const wrapper = mount(
        <IconPlugin
          config={{
            action: ctx => {
              ctx.icon.showLoading();
              setTimeout(() => {
                ctx.icon.hideLoading();
              }, 2000);
            }
          }}
          autoLoading={false}
          type="appstore"
        />
      );
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.find(Icon).at(1).simulate('click');
      jest.runAllTimers(); // 时间快进

      wrapper.update();
      wrapper.unmount();
    })

    it('autoLoading is true', () => {
      const wrapper = mount(
        <IconPlugin
          config={{
            action: [
              () => {
                console.log('next action');
              },
              'finallyAction'
            ]
          }}
          type="appstore"
        />
      );
      expect(wrapper.render()).toMatchSnapshot();
    })

    it('autoLoading is true, has lastAction.final', () => {
      const wrapper = mount(
        <IconPlugin
          config={{
            action: [
              () => {
                console.log('next action');
              },
              {
                type: jest.fn(),
                final: () => {
                  console.log('final action')
                }
              }
            ]
          }}
          type="appstore"
        />
      );
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.find(Icon).at(1).simulate('click');
    })

    it('autoLoading is true, no lastAction.final', () => {
      const wrapper = mount(
        <IconPlugin
          config={{
            action: [
              () => {
                console.log('next action');
              },
              {
                type: jest.fn(),
                before: () => {
                  console.log('final action')
                }
              }
            ]
          }}
          type="appstore"
        />
      );
      expect(wrapper.render()).toMatchSnapshot();
      wrapper.find(Icon).at(1).simulate('click');
    })

    it('no config', () => {
      const wrapper = mount(
        <IconPlugin type="appstore"/>
      );
      expect(wrapper.render()).toMatchSnapshot();
    })

  })
})
