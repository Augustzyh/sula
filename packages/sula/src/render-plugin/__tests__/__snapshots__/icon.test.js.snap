// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon Icon test basic icon 1`] = `null`;

exports[`icon Icon test has text 1`] = `
<span
  class="sula-icon-clickable"
>
  <span
    aria-label="tablet"
    class="anticon anticon-tablet"
    role="img"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="tablet"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M800 64H224c-35.3 0-64 28.7-64 64v768c0 35.3 28.7 64 64 64h576c35.3 0 64-28.7 64-64V128c0-35.3-28.7-64-64-64zM512 824c-22.1 0-40-17.9-40-40s17.9-40 40-40 40 17.9 40 40-17.9 40-40 40z"
      />
    </svg>
  </span>
  <span
    style="margin-left: 4px;"
  >
    testIcon
  </span>
</span>
`;

exports[`icon Icon test no text 1`] = `
<span
  aria-label="tablet"
  class="anticon anticon-tablet testClassName"
  role="img"
>
  <svg
    aria-hidden="true"
    class=""
    data-icon="tablet"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M800 64H224c-35.3 0-64 28.7-64 64v768c0 35.3 28.7 64 64 64h576c35.3 0 64-28.7 64-64V128c0-35.3-28.7-64-64-64zM512 824c-22.1 0-40-17.9-40-40s17.9-40 40-40 40 17.9 40 40-17.9 40-40 40z"
    />
  </svg>
</span>
`;

exports[`icon LoadingIcon basic loadingIcon 1`] = `
<span
  aria-label="loading"
  class="anticon anticon-loading sula-icon-loading"
  role="img"
>
  <svg
    aria-hidden="true"
    class="anticon-spin"
    data-icon="loading"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="0 0 1024 1024"
    width="1em"
  >
    <path
      d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
    />
  </svg>
</span>
`;

exports[`icon LoadingIconManager autoLoading is false 1`] = `
<span>
  <span
    aria-label="loading"
    class="anticon anticon-loading sula-icon-loading"
    role="img"
    style="display: none;"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="anticon-spin"
      data-icon="loading"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="0 0 1024 1024"
      width="1em"
    >
      <path
        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
      />
    </svg>
  </span>
  <span
    aria-label="appstore"
    class="anticon anticon-appstore sula-icon-clickable"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="appstore"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M464 144H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H212V212h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H612V212h200v200zM464 544H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H212V612h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H612V612h200v200z"
      />
    </svg>
  </span>
</span>
`;

exports[`icon LoadingIconManager autoLoading is true 1`] = `
<span>
  <span
    aria-label="loading"
    class="anticon anticon-loading sula-icon-loading"
    role="img"
    style="display: none;"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="anticon-spin"
      data-icon="loading"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="0 0 1024 1024"
      width="1em"
    >
      <path
        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
      />
    </svg>
  </span>
  <span
    aria-label="appstore"
    class="anticon anticon-appstore sula-icon-clickable"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="appstore"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M464 144H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H212V212h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H612V212h200v200zM464 544H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H212V612h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H612V612h200v200z"
      />
    </svg>
  </span>
</span>
`;

exports[`icon LoadingIconManager autoLoading is true, has lastAction.final 1`] = `
<span>
  <span
    aria-label="loading"
    class="anticon anticon-loading sula-icon-loading"
    role="img"
    style="display: none;"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="anticon-spin"
      data-icon="loading"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="0 0 1024 1024"
      width="1em"
    >
      <path
        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
      />
    </svg>
  </span>
  <span
    aria-label="appstore"
    class="anticon anticon-appstore sula-icon-clickable"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="appstore"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M464 144H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H212V212h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H612V212h200v200zM464 544H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H212V612h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H612V612h200v200z"
      />
    </svg>
  </span>
</span>
`;

exports[`icon LoadingIconManager autoLoading is true, no lastAction.final 1`] = `
<span>
  <span
    aria-label="loading"
    class="anticon anticon-loading sula-icon-loading"
    role="img"
    style="display: none;"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="anticon-spin"
      data-icon="loading"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="0 0 1024 1024"
      width="1em"
    >
      <path
        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
      />
    </svg>
  </span>
  <span
    aria-label="appstore"
    class="anticon anticon-appstore sula-icon-clickable"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="appstore"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M464 144H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H212V212h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H612V212h200v200zM464 544H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H212V612h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H612V612h200v200z"
      />
    </svg>
  </span>
</span>
`;

exports[`icon LoadingIconManager no config 1`] = `
<span
  aria-label="appstore"
  class="anticon anticon-appstore"
  role="img"
>
  <svg
    aria-hidden="true"
    class=""
    data-icon="appstore"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M464 144H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H212V212h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V160c0-8.8-7.2-16-16-16zm-52 268H612V212h200v200zM464 544H160c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H212V612h200v200zm452-268H560c-8.8 0-16 7.2-16 16v304c0 8.8 7.2 16 16 16h304c8.8 0 16-7.2 16-16V560c0-8.8-7.2-16-16-16zm-52 268H612V612h200v200z"
    />
  </svg>
</span>
`;
